File: python/haune/_geoip.py
"""Resolve a proxy's exit country / timezone / lat-lon / IP so the identity's geo
signals (timezone, geolocation, WebRTC IP) can be made coherent with the exit — the
same ``geoip=True`` behaviour users expect from an anti-detect browser."""
from __future__ import annotations

from dataclasses import dataclass

import requests


@dataclass
class ExitInfo:
    ip: str | None = None
    country: str | None = None
    timezone: str | None = None
    lat: float | None = None
    lon: float | None = None


def resolve_exit(proxy: str | None) -> ExitInfo | None:
    """One round-trip through the proxy to ip-api.com (free, no token)."""
    if not proxy:
        return None
    try:
        proxies = {"http": proxy, "https": proxy}
        r = requests.get(
            "http://ip-api.com/json?fields=status,countryCode,timezone,lat,lon,query",
            proxies=proxies,
            timeout=15,
        )
        j = r.json()
        if j.get("status") != "success":
            return None
        return ExitInfo(
            ip=j.get("query"),
            country=j.get("countryCode"),
            timezone=j.get("timezone"),
            lat=j.get("lat"),
            lon=j.get("lon"),
        )
    except Exception:
        return None

File: clients/dotnet/GeoIp.cs
using System.Net;
using System.Text.Json;

namespace Haune;

/// <summary>The public IP / country / timezone / lat-lon seen at a proxy's exit, so the
/// identity's geo signals can be made coherent with it (geoip=true behaviour).</summary>
public sealed record ExitInfo(string? Ip, string? Country, string? Timezone, double? Lat, double? Lon);

public static class GeoIp
{
    /// <summary>One round-trip through the proxy to ip-api.com (free, no token).</summary>
    public static async Task<ExitInfo?> ResolveAsync(ProxyInfo proxy, CancellationToken ct = default)
    {
        try
        {
            var handler = new HttpClientHandler
            {
                UseProxy = true,
                Proxy = new WebProxy(proxy.Server)
                {
                    Credentials = proxy.Username is null
                        ? null
                        : new NetworkCredential(proxy.Username, proxy.Password),
                },
            };
            using var http = new HttpClient(handler) { Timeout = TimeSpan.FromSeconds(15) };
            var json = await http.GetStringAsync(
                "http://ip-api.com/json?fields=status,countryCode,timezone,lat,lon,query", ct);
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (root.TryGetProperty("status", out var st) && st.GetString() != "success")
                return null;
            return new ExitInfo(
                root.TryGetProperty("query", out var q) ? q.GetString() : null,
                root.TryGetProperty("countryCode", out var c) ? c.GetString() : null,
                root.TryGetProperty("timezone", out var t) ? t.GetString() : null,
                root.TryGetProperty("lat", out var la) && la.ValueKind == JsonValueKind.Number ? la.GetDouble() : null,
                root.TryGetProperty("lon", out var lo) && lo.ValueKind == JsonValueKind.Number ? lo.GetDouble() : null);
        }
        catch
        {
            return null;
        }
    }
}
